README.md
<a href="https://www.gotoiot.com/">
    <img src="doc/gotoiot-logo.png" alt="logo" title="Goto IoT" align="right" width="60" height="60" />
</a>

Web App Full Stack Base
=======================

*Ayudaría mucho si apoyaras este proyecto con una ⭐ en Github!*

Este proyecto es una aplicación web fullstack que se ejecuta sobre el ecosistema `Docker`. Está compuesta por un compilador de `TypeScript` que te permite utilizar este superset de JavaScript para poder programar un `cliente web`. También tiene un servicio en `NodeJS` que te permite ejecutar código en backend y al mismo tiempo disponibilizar el código del cliente web para interactar con el servicio. Además tiene una `base de datos` MySQL que puede interactuar con el backend para guardar y consultar datos, y de manera adicional trae un `administrador` de base de datos para poder administrar la base en caso que lo necesites.

La aplicación IoT de base que viene con este proyecto se encarga de crear una tabla llamada `Devices` en la base de datos, y la idea es que vos puedas desarrollar el código de backend y frontend que te permita controlar desde el navegador el estado de los devices de un hogar inteligente - *como pueden ser luces, TVs, ventiladores, persianas, enchufes y otros* - y almacenar los estados de cada uno en la base de datos. 

Realizando estas tareas vas a a tener una aplicación fullstack IoT del mundo real que utiliza tecnologías actuales en la que un backend es capaz de interactuar con una DB para cumplir con las peticiones de control que se le mandan desde el cliente web.

En esta imagen podés ver una posible implementación del cliente web que controla los artefactos del hogar.

![architecture](doc/webapp-example-1.png)

## Comenzando 🚀

Esta sección es una guía con los pasos escenciales para que puedas poner en marcha la aplicación.

<details><summary><b>Mira los pasos necesarios</b></summary><br>

### Instalar las dependencias

Para correr este proyecto es necesario que instales `Docker` y `Docker Compose`. 

En [este artículo](https://www.gotoiot.com/pages/articles/docker_installation_linux/) publicado en nuestra web están los detalles para instalar Docker y Docker Compose en una máquina Linux. Si querés instalar ambas herramientas en una Raspberry Pi podés seguir [este artículo](https://www.gotoiot.com/pages/articles/rpi_docker_installation) de nuestra web que te muestra todos los pasos necesarios.

En caso que quieras instalar las herramientas en otra plataforma o tengas algún incoveniente, podes leer la documentación oficial de [Docker](https://docs.docker.com/get-docker/) y también la de [Docker Compose](https://docs.docker.com/compose/install/).

Continua con la descarga del código cuando tengas las dependencias instaladas y funcionando.

### Descargar el código

Para descargar el código, lo más conveniente es que realices un `fork` de este proyecto a tu cuenta personal haciendo click en [este link](https://github.com/gotoiot/app-fullstack-base/fork). Una vez que ya tengas el fork a tu cuenta, descargalo con este comando (acordate de poner tu usuario en el link):

```
git clone https://github.com/USER/app-fullstack-base.git
```

> En caso que no tengas una cuenta en Github podes clonar directamente este repo.

### Ejecutar la aplicación

Para ejecutar la aplicación tenes que correr el comando `docker-compose up` desde la raíz del proyecto. Este comando va a descargar las imágenes de Docker de node, de typescript, de la base datos y del admin de la DB, y luego ponerlas en funcionamiento. 

Para acceder al cliente web ingresa a a la URL [http://localhost:8000/](http://localhost:8000/) y para acceder al admin de la DB accedé a [localhost:8001/](http://localhost:8001/). 

Si pudiste acceder al cliente web y al administrador significa que la aplicación se encuentra corriendo bien. 

> Si te aparece un error la primera vez que corres la app, detené el proceso y volvé a iniciarla. Esto es debido a que el backend espera que la DB esté creada al iniciar, y en la primera ejecución puede no alcanzar a crearse. A partir de la segunda vez el problema queda solucionado.

</details>

Continuá explorando el proyecto una vez que lo tengas funcionando.

## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.

<details><summary><b>Lee cómo configurar la aplicación</b></summary><br>

### Configuración de la DB

Como ya comprobaste, para acceder PHPMyAdmin tenés que ingresar en la URL [localhost:8001/](http://localhost:8001/). En el login del administrador, el usuario para acceder a la db es `root` y contraseña es la variable `MYSQL_ROOT_PASSWORD` del archivo `docker-compose.yml`.

Para el caso del servicio de NodeJS que se comunica con la DB fijate que en el archivo `src/backend/mysql-connector.js` están los datos de acceso para ingresar a la base.

Si quisieras cambiar la contraseña, puertos, hostname u otras configuraciones de la DB deberías primero modificar el servicio de la DB en el archivo `docker-compose.yml` y luego actualizar las configuraciones para acceder desde PHPMyAdmin y el servicio de NodeJS.

### Estructura de la DB

Al iniciar el servicio de la base de datos, si esta no está creada toma el archivo que se encuentra en `db/dumps/smart_home.sql` para crear la base de datos automáticamente.

En ese archivo está la configuración de la tabla `Devices` y otras configuraciones más. Si quisieras cambiar algunas configuraciones deberías modificar este archivo y crear nuevamente la base de datos para que se tomen en cuenta los cambios.

Tené en cuenta que la base de datos se crea con permisos de superusuario por lo que no podrías borrar el directorio con tu usuario de sistema, para eso debés hacerlo con permisos de administrador. En ese caso podés ejecutar el comando `sudo rm -r db/data` para borrar el directorio completo.

</details>


## Detalles principales 🔍

En esta sección vas a encontrar las características más relevantes del proyecto.

<details><summary><b>Mira los detalles más importantes de la aplicación</b></summary><br>
<br>

### Arquitectura de la aplicación

Como ya pudiste ver, la aplicación se ejecuta sobre el ecosistema Docker, y en esta imagen podés ver el diagrama de arquitectura.

![architecture](doc/architecture.png)

### El cliente web

El cliente web es una Single Page Application que se comunica con el servicio en NodeJS mediante JSON a través de requests HTTP. Puede consultar el estado de dispositivos en la base de datos (por medio del servicio en NodeJS) y también cambiar el estado de los mismos. Los estilos del código están basados en **Material Design**.

### El servicio web

El servicio en **NodeJS** posee distintos endpoints para comunicarse con el cliente web mediante requests HTTP enviando **JSON** en cada transacción. Procesando estos requests es capaz de comunicarse con la base de datos para consultar y controlar el estado de los dispositivos, y devolverle una respuesta al cliente web también en formato JSON. Así mismo el servicio es capaz de servir el código del cliente web.

### La base de datos

La base de datos se comunica con el servicio de NodeJS y permite almacenar el estado de los dispositivos en la tabla **Devices**. Ejecuta un motor **MySQL versión 5.7** y permite que la comunicación con sus clientes pueda realizarse usando usuario y contraseña en texto plano. En versiones posteriores es necesario brindar claves de acceso, por este motivo la versión 5.7 es bastante utilizada para fases de desarrollo.

### El administrador de la DB

Para esta aplicación se usa **PHPMyAdmin**, que es un administrador de base de datos web muy utilizado y que podés utilizar en caso que quieras realizar operaciones con la base, como crear tablas, modificar columnas, hacer consultas y otras cosas más.

### El compilador de TypeScript

**TypeScript** es un lenguaje de programación libre y de código abierto desarrollado y mantenido por Microsoft. Es un superconjunto de JavaScript, que esencialmente añade tipos estáticos y objetos basados en clases. Para esta aplicación se usa un compilador de TypeScript basado en una imagen de [Harmish](https://hub.docker.com/r/harmish) en Dockerhub, y está configurado para monitorear en tiempo real los cambios que se realizan sobre el directorio **src/frontend/ts** y automáticamente generar código compilado a JavaScript en el directorio  **src/frontend/js**. Los mensajes del compilador aparecen automáticamente en la terminal al ejecutar el comando **docker-compose up**.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas. Los detalles sobre cómo funcionan los servicios los podés ver directamente en el archivo **docker-compose.yml**.

### Organización del proyecto

En la siguiente ilustración podés ver cómo está organizado el proyecto para que tengas en claro qué cosas hay en cada lugar.

```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

> No olvides ir poniendo tus cambios en el archivo `CHANGELOG.md` a medida que avanzas en el proyecto.

</details>

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código.

<details><summary><b>Mira los detalles de implementación</b></summary><br>

### Frontend 👨🏻‍💻

Este proyecto es una Single Page Application (SPA) basada en el template del repositorio de [mramos88](https://github.com/mramos88/app-fullstack-base-2023-i08). La aplicación ha sido modificada para permitir el alta, baja, modificación y control de dispositivos (de tres tipos: lamparas, cotinas y aires acondicionados), así como el inicio de sesión de usuarios mediante correo electrónico y contraseña.

La aplicación envía los estados de los dispositivos al servidor y los almacena en una base de datos cada vez que ocurre un cambio. Todos los datos relevantes de los dispositivos, como el ID, nombre, descripción, tipo y estado, también se almacenan en la base de datos.

Los usuarios pueden acceder a sus cuentas utilizando sus credenciales de inicio de sesión. Estos datos también se almacenan en la base de datos para su posterior autenticación.

La SPA proporciona una experiencia de usuario fluida y receptiva, permitiendo a los usuarios realizar todas las acciones necesarias para administrar los dispositivos.

¡Explora todas las funcionalidades que ofrece esta aplicación!

![devices](doc/devices.png)

#### Inicio de sesión de usuarios 👥

Se ha implementado la opción de inicio de sesión utilizando el correo electrónico y la contraseña. En esta etapa, se han agregado dos usuarios para testing: el usuario número 1 es `admin@admin.com` con la contraseña `secreto`, y el segundo usuario es `user@user.com` con la contraseña `1234`. queda como mejora a implementar la funcionalidad de alta, baja y modificación de usuarios.

El correo electrónico del usuario con el que se ha iniciado sesión se muestra en la pantalla de dispositivos. Por el momento, al iniciar sesión con cualquiera de los usuarios, solo se muestran los dispositivos del usuario `admin@admin.com`:

![login](doc/login.png)

Si no se ingresa un mail, la aplicación informará de dicho error:

![login_failed1](doc/login_failed1.png)

De misma manera si no se ingresa la contraseña:

![login_failed2](doc/login_failed2.png)

En el caso que las credenciales sea incorrectas se informará mediante un aviso:

![login_failed3](doc/login_failed3.png)

> 👀 ¡Importante!: Al levantar el `docker-compose`, en ocasiones, se ha observado que no conecta a la DB y por esto recibirás un mensaje de `Error de usuario y/o contraseña`. Si esto ocurre detené el proceso y volvé a iniciar la app. Si continua prueba actualizando el `index.js`.

#### Agregar un dispositivo ➕

Para agregar un dispositivo, simplemente haz clic en el botón `+` en la parte inferior de la página y completa los datos del formulario. Luego, confirma la acción haciendo clic en el botón `Agregar` o `cancelar` si no deseas agregar un nuevo dispositivo. Puedes agregar tres tipos de dispositivos: Lampara (tipo de dispositivo 1), Cortina (tipo de dispositivo 2) y Aire (tipo de dispositivo 3).

- Lampara: Tiene un interruptor de encendido y apagado.
- Cortina: Tiene una barra deslizadora para ajustar el valor entre 0% y 100%, con intervalos de 10%.
- Aire Acondicionado: Tiene una barra deslizadora para ajustar la temperatura entre 16°C y 30°C, con intervalos de 1°C.

![add](doc/add.png)

La aplicación confirmará que el dispositivo ha sido agregado:

![added](doc/added.png)

#### Eliminar un dispositivo 🗑️

En cada uno de los dispositivos listados, hay un botón `Eliminar`. Al hacer clic en él, se mostrará un menú para confirmar la acción:

![erase](doc/erase.png)

De manera similar al agregar dispositivos, recibirás una notificación informando que el dispositivo ha sido eliminado:

![erased](doc/erased.png)

#### Modificar un dispositivo 📝

En cada uno de los dispositivos listados, hay un botón `Modificar`. Al hacer clic en él, se mostrará un menú para confirmar la acción. En esta instancia, podrás editar los mismos atributos que se solicitan al agregar un dispositivo:

![modify](doc/modify.png)

Luego, recibirás una notificación confirmando que la modificación se ha realizado con éxito:

![modified](doc/modified.png)

#### Refrescar lista de dispositivos 🔄

En la parte inferior, junto al botón `Agregar` dispositivos, encontrarás un botón `Refrescar`. Este botón es útil para actualizar los valores de los dispositivos.


### Backend 👨🏻‍💻

El backend proporciona una API con endpoints para realizar diversas funciones relacionadas con dispositivos y usuarios. La aplicación utiliza una base de datos para almacenar la información. A continuación, se detallan mas informacion sobre la estructura de la base de datos y endpoints disponibles:

#### Base de datos 💾:

1. Users:
    *   id: almacena el ID único de cada usuario.
    *   email: almacena la dirección de correo electrónico del usuario.
    *   password: almacena la contraseña del usuario.

> 👀 ¡Recuerda!: No se deben almacenar contraseñas directamente en la base de datos. Se recomienda utilizar técnicas de encriptación, como el hash y salt, para almacenar y verificar las contraseñas de forma segura.

2. Devices:
    *   id: almacena el ID único de cada dispositivo.
    *   name: almacena el nombre del dispositivo.
    *   description: almacena la descripción del dispositivo.
    *   state: almacena el estado actual del dispositivo.
    *   type: almacena el tipo de dispositivo (1 para Lampara, 2 para cortina, 3 para aire acondicionado).
    *   user_id: almacena el ID del usuario al que pertenece el dispositivo.

#### Endpoints ⚙️ :
A continuación se presentan los endpoints disponibles junto con sus descripciones:

<details><summary><b>Ver los endpoints disponibles</b></summary><br>

1. Obtener toda la lista de dispositivos y sus características de la base de datos:

    *   URL: http://localhost:8000/devices
    *   Método: GET
    *   Body: Ninguno
    *   Respuesta:
        *   Código de respuesta 200 - OK, junto con un JSON que contiene los dispositivos.
        *   Código de respuesta 400 - Error.
        
    Ejemplo de respuesta exitosa (código 200):
    ```json
        [
            {
                "id": 1,
                "name": "Lampara",
                "description": "Living",
                "state": 0,
                "type": 1,
                "user_id": 1
            },
            {
                "id": 2,
                "name": "Cortina",
                "description": "Habitacion",
                "state": 60,
                "type": 2,
                "user_id": 1
            },
            {
                "id": 3,
                "name": "Aire",
                "description": "Living",
                "state": 24,
                "type": 3,
                "user_id": 1
            },
            {
                "id": 4,
                "name": "Lampara",
                "description": "Cocina",
                "state": 1,
                "type": 1,
                "user_id": 2
            },
            {
                "id": 5,
                "name": "Cortina",
                "description": "Comedor",
                "state": 20,
                "type": 2,
                "user_id": 2
            },
            {
                "id": 6,
                "name": "Aire",
                "description": "Comedor",
                "state": 20,
                "type": 3,
                "user_id": 2
            }
        ]
    ```

2. Buscar un dispositivo en particular por medio de su id:
    *   URL: http://localhost:8000/devices/id
    *   Método: GET
    *   Body: Ninguno
    *   Respuesta:
        *   Código de respuesta 200 - OK, junto con un JSON que contiene el dispositivo solicitado.
        *   Código de respuesta 400 - Error.
        Ejemplo de respuesta exitosa (código 200):
        ```json
        [
            {
                "id": 4,
                "name": "Lampara",
                "description": "Cocina",
                "state": 1,
                "type": 1,
                "user_id": 2
            },
        ]
        ```
3. Obtener un dispositivo por su id:
    *   URL: http://localhost:8000/devices/:id
    *   Método: GET
    *   Body: Ninguno
    *   Respuesta:
        *   Código de respuesta 200 - OK, junto con un JSON que contiene el dispositivo solicitado.
        *   Código de respuesta 400 - Error.

4. Actualizar el estado de un dispositivo:

CHANGELOG.md
# DAW Base App - Changes Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).


## 2.3.0

* Project modification
    * Allows login by email and password
    * Adds the feature of adding, removing or editing devices
    * Adds a third type of device, air conditioning
    * Incorporates visual improvements of the buttons
    * Adds 'toast' type messages
    * Adds new info to README accordingly
    * Changes webapp example image


## 2.2.0

* Project modification
    * Adds TypeScript compiler service to Docker Compose
    * Reestructures frontend folder for TypeScript
    * Adds new info to README accordingly
    * Changes project architecture image

## 2.1.0

* Project modification
    * Enhaces README accordingly to Goto IoT
    * Adds example of finished application
    * Removes unnecessary frontend images
    * Changes src code folders names

## 2.0.0

* Project modification
    * Changes project and organization names
    * Removes Typescript container
    * Removes Typescript Code
    * Executes Javascript code directly
    * Changes licence to MIT
    * Modifies README accordingly

## 1.0.0

* Project creation
    * Docker Compose implementation for whole project.
    * Typescript compilation into docker-compose.
    * MySQL 5.7 DB Server.
    * PHPMyAdmin.
    * NodeJS backend application.
    * Materialize CSS framework.


src/frontend/js/main.js
var M;
class Main {
    constructor() {
        this.framework = new Framework();
    }
    // Función para realizar la consulta inicial
    initialQuery() {
        this.framework.devRequest("GET", "http://localhost:8000/devices", this);
    }
    // Función para consultar un dispositivo específico
    queryDevice(devID, action) {
        return this.framework.makeRequest("GET", `http://localhost:8000/devices/${devID}`, this, action);
    }
    // Función para agregar un dispositivo nuevo
    addDevice(name, description, type, user_id) {
        let state = 0;
        if (type === "3") {
            state = 24;
        }
        const newDevice = { name, description, type, state, user_id };
        this.framework.addRequest("POST", "http://localhost:8000/devices/", this, newDevice);
    }
    // Función para cambiar el estado de un dispositivo
    changeState(devID, state) {
        const changedDevice = { state };
        this.framework.changeRequest("PUT", `http://localhost:8000/devices/state/${devID}`, this, changedDevice);
    }
    // Función para cambiar los detalles de un dispositivo
    changeDevice(devID, name, description, type) {
        const changedDevice = { name, description, type };
        this.framework.changeRequest("PUT", `http://localhost:8000/devices/${devID}`, this, changedDevice);
    }
    // Función para eliminar un dispositivo
    deleteDevice(devID) {
        this.framework.deleteRequest("DELETE", `http://localhost:8000/devices/${devID}`, this);
    }
    // Función para verificar usuarios
    loginUser(email, password) {
        const user = { email, password };
        return this.framework.loginRequest("POST", `http://localhost:8000/login`, this, user);
    }
    // Función para obtener el user_id --- pendiente implementacion
    currentUserID(email) {
        this.framework.currentUserIDRequest("GET", "http://localhost:8000/user_id", this, email);
    }
    // Función para actualizar la SPA (Single Page Application)
    refreshSPA(tipoModal) {
        this.initialQuery();
        const modal = document.getElementById(tipoModal);
        const instanceModal = M.Modal.getInstance(modal);
        instanceModal.close();
    }
    // Función refresh dispositivos
    refreshDevices() {
        this.initialQuery();
    }
    loadGrid(devList) {
        const devbox = document.getElementById("devbox");
        let grid = `<ul class="collection">`;
        // Ocultar elementos relacionados al inicio de sesión
        const divEmail = document.getElementById("divEmail");
        divEmail.hidden = true;
        const divPassword = document.getElementById("divPassword");
        divPassword.hidden = true;
        const divEnter = document.getElementById("divEnter");
        divEnter.hidden = true;
        const welcome = document.getElementById("welcome");
        welcome.hidden = true;
        // Mostrar el título y el email del usuario
        const title = document.getElementById("title");
        title.hidden = false;
        const nombre = document.getElementById("email").value;
        const hello = document.getElementById("hello");
        hello.innerHTML = `Usuario: ${nombre}`;
        hello.hidden = false;
        // Mostrar el botón para agregar un dispositivo
        const devbtnadd = document.getElementById("divAdd");
        devbtnadd.hidden = false;
        // Mostrar el botón logout
        const logbtn = document.getElementById("logout-button");
        logbtn.hidden = false;
        // Botón refresh
        const refreshButton = document.getElementById("refresh-button");
        refreshButton.hidden = false;
        refreshButton.addEventListener("click", () => {
            this.refreshDevices();
        });
        // Filtrar los dispositivos del usuario "1" admin@admin.com (se puede cambiar por "2" user@user.com), para el ejemplo es admin@admin.com. Queda pendiente implementacion para que traiga las vistas de dispositivos dependiendo del usuario que se loguea
        const currentUserDevices = devList.filter(dev => dev.user_id === 1);
        for (const dev of currentUserDevices) {
            let deviceType = "";
            let deviceImage = "";
            let extraFields = "";
            // Comprobar el tipo de dispositivo y construir la grid correspondiente
            if (dev.type === 1) {
                deviceType = "lightbulb";
                deviceImage = "static/images/lightbulb.png";
                extraFields = `
            <div class="switch">
              <label>
                Off
                <input id="value${dev.id}" type="checkbox" ${dev.state ? 'checked' : ''}>
                <span class="lever"></span>
                On
              </label>
            </div>`;
            }
            else if (dev.type === 2) {
                deviceType = "window";
                deviceImage = "static/images/window.png";
                extraFields = `
            <form action="#">
              <p class="range-field">
                <input type="range" id="value${dev.id}" min="0" max="100" step="10" value="${dev.state}">
                <label for="value${dev.id}">${dev.state} %</label>
              </p>
            </form>`;
            }
            else if (dev.type === 3) {
                deviceType = "air";
                deviceImage = "static/images/air.png";
                extraFields = `
            <form action="#">
              <p class="range-field">
                <input type="range" id="value${dev.id}" min="14" max="30" step="1" value="${dev.state}">
                <label for="value${dev.id}">${dev.state} °C</label>
              </p>
            </form>`;
            }
            grid += `
          <li class="collection-item avatar">
            <img src="${deviceImage}" alt=" " class="circle">
            <span class="title negrita" id="${dev.name}">${dev.name}</span>
            <p>${dev.description}</p>
            <a class="secondary-content">
              ${extraFields}
            </a>
            <br>
            <button class="btn waves-effect waves-light button-view" id="mod-button${dev.id}" name="hello_button">
              <i id="mod-button${dev.id}" class="material-icons left">edit</i>Modificar
            </button>
            <button class="btn waves-effect waves-light button-view" id="del-button${dev.id}" name="hello_button">
              <i id="del-button${dev.id}" class="material-icons left">delete</i>Eliminar
            </button>
          </li>`;
        }
        grid += `</ul>`;
        devbox.innerHTML = grid;
        // Asignar eventos a los elementos creados dinámicamente
        for (let dev of currentUserDevices) {
            document.getElementById("value" + dev.id).addEventListener("click", this);
            document.getElementById("mod-button" + dev.id).addEventListener("click", this);
            document.getElementById("del-button" + dev.id).addEventListener("click", this);
        }
    }
    loadDevice(device, action) {
        // Comprobar la acción y asignar los valores correspondientes a los elementos del formulario de edición o eliminación
        if (action == "edit") {
            document.getElementById("edit-device-id").value = device.id.toString();
            document.getElementById("edit-name").value = device.name;
            document.getElementById("edit-description").value = device.description;
            let select = document.getElementById("select-edit-type");
            var instanceSelect = M.FormSelect.getInstance(select);
            select.value = device.type.toString();
            instanceSelect.destroy();
            M.FormSelect.init(select, "");
        }
        else if (action == "delete") {
            document.getElementById("delete-id").innerHTML = device.id.toString();
            document.getElementById("delete-name").innerHTML = device.name;
            document.getElementById("delete-description").innerHTML = device.description;
            document.getElementById("input-delete-id").value = device.id.toString();
        }
    }
    openModal(mode) {
        // Abrir el modal correspondiente según el modo especificado
        let modal = document.getElementById(mode);
        var instanceModal = M.Modal.getInstance(modal);
        instanceModal.open();
    }
    closeModal(mode) {
        // Cerrar el modal correspondiente según el modo especificado
        let modal = document.getElementById(mode);
        var instanceModal = M.Modal.getInstance(modal);
        instanceModal.close();
    }
    handleEvent(object) {
        let objEvent;
        objEvent = object.target;
        console.log(objEvent);
        let devstate;
        // Manejar eventos según el elemento que haya sido activado
        if (objEvent.id == "enter-button") {
            const email = document.getElementById("email").value;
            if (email.length === 0) {
                M.toast({ html: '📧 Ingrese email', classes: 'toast-centered' });
            }
            else {
                const password = document.getElementById("password").value;
                if (password.length === 0) {
                    M.toast({ html: '🔑 Ingrese contraseña', classes: 'toast-centered' });
                }
                else {
                    this.loginUser(email, password).then((loginOK) => {
                        if (loginOK) {
                            M.toast({ html: "🚪 Ingresando..." });
                            this.initialQuery();
                        }
                        else {
                            M.toast({ html: "❌ Error de usuario y/o contraseña", classes: 'toast-centered' });
                        }
                    })
                        .catch((error) => {
                        M.toast({ html: "🤔 Error desconocido", classes: 'toast-centered' });
                    });
                }
            }
        }
        else if (objEvent.type == "checkbox") {
            // Cambiar el estado del dispositivo según el estado del checkbox
            if (objEvent.checked == true) {
                devstate = 1;
            }
            else {
                devstate = 0;
            }
            let devid = parseInt(objEvent.id.replace('value', ''));
            this.changeState(devid, devstate);
        }
        else if (objEvent.type == "range") {
            // Cambiar el estado del dispositivo según el valor del rango
            devstate = parseInt(objEvent.value);
            let devid = parseInt(objEvent.id.replace('value', ''));
            this.changeState(devid, devstate);
        }
        else if (objEvent.id.startsWith("mod-button")) {
            // Obtener el ID del dispositivo para editar y abrir el modal de edición
            let devid = parseInt(objEvent.id.replace('mod-button', ''));
            this.queryDevice(devid, "edit");
            this.openModal("edit-modal");
        }
        else if (objEvent.id.startsWith("del-button")) {
            // Obtener el ID del dispositivo para eliminar y abrir el modal de eliminación
            let devid = parseInt(objEvent.id.replace('del-button', ''));
            this.queryDevice(devid, "delete");
            this.openModal("delete-modal");
        }
        else if (objEvent.id == "add-button") {
            // Obtener el ID del dispositivo para agregar y abrir el modal de agregar
            let devid = parseInt(objEvent.id);
            this.queryDevice(devid, "add");
            this.openModal("add-modal");
        }
        else if (objEvent.id == "cancel-edit") {
            // Cerrar el modal de edición
            this.closeModal("edit-modal");
        }
        else if (objEvent.id == "confirm-edit") {
            // Obtener los valores editados del formulario y aplicar los cambios al dispositivo
            let devid = +document.getElementById("edit-device-id").value;
            let name = document.getElementById("edit-name").value;
            let description = document.getElementById("edit-description").value;
            let type = parseInt(document.getElementById("select-edit-type").value);
            if (devid && name && description && type) {
                this.changeDevice(devid, name, description, type);
                M.toast({ html: '👍 Dispositivo modificado', classes: 'toast-centered' });
                this.refreshSPA("edit-modal");
            }
            else {
                M.toast({ html: '👁️‍🗨️ Completar todos los atributos!', classes: 'toast-centered' });
            }
        }
        else if (objEvent.id == "cancel-delete") {
            // Cerrar el modal de eliminación
            this.closeModal("delete-modal");
        }
        else if (objEvent.id == "confirm-delete") {
            // Obtener el ID del dispositivo a eliminar y eliminarlo
            let devid = +document.getElementById("input-delete-id").value;
            this.deleteDevice(devid);
            M.toast({ html: '🗑️ Dispositivo eliminado', classes: 'toast-centered' });
            this.refreshSPA("delete-modal");
        }
        else if (objEvent.id == "cancel-add") {
            // Cerrar el modal de agregar
            this.closeModal("add-modal");
        }
        else if (objEvent.id == "confirm-add") {
            // Obtener los valores del formulario y agregar un nuevo dispositivo
            let name = document.getElementById("txt-name").value;
            let description = document.getElementById("txt-description").value;
            let type = document.getElementById("select-type").value;
            if (name && description && type) {
                const user_id = '1'; // Agrega dispositivos del usuario "1" admin@admin.com (se puede cambiar por "2" user@user.com), para el ejemplo es admin@admin.com. Queda pendiente implementacion para cargue los dispositivos dependiendo del usuario que se loguea
                this.addDevice(name, description, type, user_id);
                M.toast({ html: '👍 Dispositivo agregado', classes: 'toast-centered' });
                this.refreshSPA("add-modal");
            }
            else {
                M.toast({ html: '👁️‍🗨️ Completar todos los atributos!', classes: 'toast-centered' });
            }
        }
    }
}
window.addEventListener("load", () => {
    let main = new Main();
    // Seleccionar todos los elementos 'select' y aplicar la inicialización de Materialize Select
    var elements = document.querySelectorAll('select');
    var instances = M.FormSelect.init(elements, "");
    // Seleccionar todos los elementos con la clase 'modal' y aplicar la inicialización de Materialize Modal
    var elementsM = document.querySelectorAll('.modal');
    M.Modal.init(elementsM, "");
    // Asignar event listeners a los botones principales
    document.getElementById("add-button").addEventListener("click", main);
    document.getElementById("enter-button").addEventListener("click", main);
    // Asignar event listeners a los botones de los modales
    document.getElementById("cancel-add").addEventListener("click", main);
    document.getElementById("confirm-edit").addEventListener("click", main);
    document.getElementById("cancel-edit").addEventListener("click", main);
    document.getElementById("confirm-delete").addEventListener("click", main);
    document.getElementById("cancel-delete").addEventListener("click", main);
    document.getElementById("confirm-add").addEventListener("click", main);
});
